File: SelectionSort/Python/selection_sort.py
"""
------ Selection Sort --------
in this we try to sort the array by select the smallest element
and put in new array and reapeat the same for all the array and compare 
the Big(O) is O(n^2)

==> how the approach works
    - first make function get me smallest index in array
    - use this to take the smallest element and put it in new array
"""
def smallest_index(arr):
    smallest_element = arr[0]
    smallest_index=0
    for i in range(1,len(arr)):
        if arr[i]<smallest_element:
            smallest_element = arr[i]
            smallest_index=i
    return smallest_index

def selection_sort(arr):
    #to make the original arr not changed 
    copy_list=arr.copy()
    sorted_arr = []
    for i in range(len(copy_list)):
        smallest_idx = smallest_index(copy_list)
        sorted_arr.append(copy_list.pop(smallest_idx))
    return sorted_arr

arr=[5,69,35,68,6,30,4]
sorted_arr=selection_sort(arr)

print(f"Original : {arr} \nSorted : {sorted_arr}")

File: SelectionSort/Typescript/SelectionSort.ts
function smallest_index(arr: number[]): number {
    let smallest_element = arr[0];
    let smallest_index = 0;
  
    // Find the index of the smallest element in the array
    for (let i = 1; i < arr.length; i++) {
      if (arr[i] < smallest_element) {
        smallest_element = arr[i];
        smallest_index = i;
      }
    }
  
    return smallest_index;
  }
  
  function selection_sort(arr: number[]): number[] {
    let copy_arr=arr.slice();
    let sorted_arr: number[] = [];
    for (let i = 0; i < arr.length; i++) {
      let smallest_idx = smallest_index(copy_arr);
      sorted_arr.push(copy_arr.splice(smallest_idx,1)[0]); 
    }
  
    console.log(sorted_arr);
    return sorted_arr;
  }
  
  let arr = [555, 69,9];
  console.log(selection_sort(arr));
  